use active: :once to avoid crashing the vm if the data arrive faster …

xadhoom opened the pull request:
…than processing capabilities

xadhoom committed:
use active: :once to avoid crashing the vm if the data arrive faster than processing capabilities

marcelog commented:
Hello,

Thanks for the contribution!

The change makes sense, since using `{:active, :true}` does not provide flow control (my bad). A few things:
- The patch is not building, could you check that out? (The missing variable can be found in the `state` variable).
- Could you post a capture of the vm crash?
- Have you tried this change in your (I assume) high load scenario? I'm interested in knowing how well it behaves and also if it correctly detects a disconnection or closing socket from the freeswitch node.

Cheers :)

xadhoom commented:
Hi,
- sorry for the buld error, I rewrote the fix on another computer, forgetting the socket variable.
  Fixed and pushed. Now all builds are ok
- Sure, but it was mainly caused by the oom killer. Basically, the VM continued to allocate messages while receiving traffic, until the oom kicked in to stop the process. Sometimes that kill resulted into a erlang crash with an "stack smashing detected" glibc error. Other times the process just died. The erl crash dump can be found here: https://gist.github.com/xadhoom/69ab37e023f4cbdae7bc
- Sure, it works ok and it does not allocate memory until killed like before. I've tried to close the sender connection and is detected correctly by the app, something like:
  15:15:41.932 [info]  Connection closed
  15:15:41.933 [info]  Terminating with :normal

The only "problem" right now is that is not very fast compared with other languages (I currently process esl events with python/twisted, but I'm searching a fastest solution. Node.js right now wins and I don't like it). But this discussion is outside the scope of this pull request :)

marcelog commented:
Thank you for the detailed information and patch :)

I'm interested in improving it. Perhaps you can send me an email to marcelog at gmail dot com so we can discuss the options? Maybe it's the elixir version, or perhaps we have to move from an "active" socket to a polling behavior, or maybe even change how the packets are processed. I'm confident that Erlang and Elixir should at least match what node offers.

Best!
